precompiles/utils/src/precompile_set.rs

```diff
@@ -18,81 +18,236 @@
+		true
+	}
+}
 
-/// Is DELEGATECALL allowed to use for a precompile.
-pub trait DelegateCallSupport: sealed::Sealed {
-	fn allow_delegate_call() -> bool;
+pub struct OnlyFrom<T>(PhantomData<T>);
+impl<T: Get<H160>> SelectorFilter for OnlyFrom<T> {
+	fn is_allowed(caller: H160, _selector: Option<u32>) -> bool {
+		caller == T::get()
+	}
 }
 
-/// DELEGATECALL is forbiden.
-pub struct ForbidDelegateCall;
-impl sealed::Sealed for ForbidDelegateCall {}
-impl DelegateCallSupport for ForbidDelegateCall {
+pub struct CallableByContract<T = ForAllSelectors>(PhantomData<T>);
+
+impl<T: SelectorFilter> PrecompileChecks for CallableByContract<T> {
 	#[inline(always)]
-	fn allow_delegate_call() -> bool {
-		false
+	fn callable_by_smart_contract(caller: H160, called_selector: Option<u32>) -> Option<bool> {
+		Some(T::is_allowed(caller, called_selector))
 	}
 }
 
-/// DELEGATECALL is allowed.
-/// Should be used with care if the precompile use
-/// custom storage, as the caller could impersonate its own caller.
-pub struct AllowDelegateCall;
-impl sealed::Sealed for AllowDelegateCall {}
-impl DelegateCallSupport for AllowDelegateCall {
+/// Precompiles are allowed to call this precompile.
+pub struct CallableByPrecompile<T = ForAllSelectors>(PhantomData<T>);
+
+impl<T: SelectorFilter> PrecompileChecks for CallableByPrecompile<T> {
 	#[inline(always)]
-	fn allow_delegate_call() -> bool {
-		true
+	fn callable_by_precompile(caller: H160, called_selector: Option<u32>) -> Option<bool> {
+		Some(T::is_allowed(caller, called_selector))
 	}
 }
 
+/// Common checks for precompile and precompile sets.
+/// Don't contain recursion check as precompile sets have recursion check for each member.
+fn common_checks<R: pallet_evm::Config, C: PrecompileChecks>(
+	handle: &mut impl PrecompileHandle,
+) -> Result<(), fp_evm::PrecompileFailure> {
+	let code_address = handle.code_address();
+	let caller = handle.context().caller;
+
+	// Check DELEGATECALL config.
+	let allow_delegate_call = C::accept_delegate_call().unwrap_or(false);
+	if !allow_delegate_call && code_address != handle.context().address {
+		return Err(revert("Cannot be called with DELEGATECALL or CALLCODE"));
+	}
+
+	// Extract which selector is called.
+	let selector = handle.input().get(0..4).map(|bytes| {
+		let mut buffer = [0u8; 4];
+		buffer.copy_from_slice(bytes);
+		u32::from_be_bytes(buffer)
+	});
+
+	// Is this selector callable from a smart contract?
+	let callable_by_smart_contract =
+		C::callable_by_smart_contract(caller, selector).unwrap_or(false);
+	if !callable_by_smart_contract {
+		handle.record_cost(RuntimeHelper::<R>::db_read_gas_cost())?;
+		let caller_code = pallet_evm::Pallet::<R>::account_codes(caller);
+		// Check that caller is not a smart contract s.t. no code is inserted into
+		// pallet_evm::AccountCodes except if the caller is another precompile i.e. CallPermit
+		if !(caller_code.is_empty() || &caller_code == &[0x60, 0x00, 0x60, 0x00, 0xfd]) {
```

Comment:
```suggestion
		if caller_has_code && (!callable_by_precompile || !caller_is_precompile) {
```

I suggest to refactor the logic here by doing the precompile check first

Reply:
Note that was don't have a `caller_is_precompile` variable yet. This value is computed only if being called by a precompile is forbidden, and calling `is_precompile` on the precompile set can perform many storage read (think XC20 precompile sets). It thus seems more performant to read the code and ensure it is either empty or the dummy code. However if the allocation of a big vec is your concern, maybe there is a way to read only the 5 first bytes of the code and ensure that the code is 5 bytes long AND those bytes match the dummy code?

Reply:
> if the allocation of a big vec is your concern, maybe there is a way to read only the 5 first bytes of the code and ensure that the code is 5 bytes long AND those bytes match the dummy code

yes it's my concern, It seems that the api not allow to do that directly. But we can verify the code len, and only if it's five, get all code bytes to check if that match the dummy code

Reply:
Exactly, I committed a change with that idea :)